### PR TITLE
Split 'npm install' and 'npm run build' stages

### LINKED ISSUE #3370: Update nodejs Dockerfile to improve layer caching
The default `Dockerfile` for a nodejs function created via `func init` is currently:

```Dockerfile
# To enable ssh & remote debugging on app service change the base image to the one below
# FROM mcr.microsoft.com/azure-functions/node:4-node16-appservice
FROM mcr.microsoft.com/azure-functions/node:4-node16

ENV AzureWebJobsScriptRoot=/home/site/wwwroot \
    AzureFunctionsJobHost__Logging__Console__IsEnabled=true

COPY . /home/site/wwwroot

RUN cd /home/site/wwwroot && \
    npm install && \
    npm run build
```

This copies the source content and runs `npm install` and `npm run build` as a single build step.

We can leverage layer caching to speed up repeated image builds by splitting this into two steps:
1. Copy `package.json` and run `npm install`
2. Copy full source and run `npm run build`

By doing this, the layer from the first step is only invalidated when `package.json` changes, so making other changes to your code doesn't invalidate this and subsequent builds will skip the `npm install` step. This speeds up repeated builds of the image and also repeated image push/pulls (since only the changed layers need transferring)

### PR DESCRIPTION
**Pull request by stuartleeks**
### Issue describing the changes in this PR

resolves #3370 

### Pull request checklist

* [X] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [X] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [X] I have added all required tests (Unit tests, E2E tests) - I don't believe there are any new tests required by this change

### CODE REVIEW

**Comment by stuartleeks** [comment]
Is there any feedback on this?

**Review by mattchenderson — DISMISSED** [review]
This is a good improvement, and these changes seem good to me. My approval should not be considered fully sufficient for a merge, but I think we should be able to take this one as-is. We should still do some quick validation passes though first.

**Comment by mattchenderson** [comment]
@ejizba I'd like to also make sure you're aware of these changes before we complete this. Could you please take a look?

File: src/Azure.Functions.Cli/StaticResources/Dockerfile.node18 [hunk 1]
```diff
@@ -5,7 +5,12 @@ FROM mcr.microsoft.com/azure-functions/node:4-node18
...
-RUN cd /home/site/wwwroot && \
-    npm install
\ No newline at end of file
+# Copy package.json and install packages before copying the rest of the code to enable caching
+COPY package.json package.json
+RUN npm install
+
+# Copy the rest of the code
+COPY . .
+RUN  npm run build
```
Comment (ejizba) [inline src/Azure.Functions.Cli/StaticResources/Dockerfile.node18:16]: Hi @stuartleeks sorry for the delay and thanks for the contribution! Please remove the `npm run build` step from the 3 files without `.typescript` in the name. These 3 files should be for JavaScript, which doesn't need a build step and didn't have one before your change

**Review by ejizba — CHANGES_REQUESTED** [review]

File: src/Azure.Functions.Cli/StaticResources/Dockerfile.node18 (hunk 1 above)
Comment (stuartleeks) [reply src/Azure.Functions.Cli/StaticResources/Dockerfile.node18]: Oops - good catch, thanks!

Comment (stuartleeks) [reply src/Azure.Functions.Cli/StaticResources/Dockerfile.node18]: Updated and rebased on latest `v4.x`

**Review by ejizba — APPROVED** [review]
Thanks!

**Pull request merged**